lib/msal-common/src/cache/CacheManager.ts

```diff
@@ -204,23 +214,103 @@ export abstract class CacheManager implements ICacheManager {
+        const tenantProfiles = new Map<String, TenantProfile>();
+        const accountId = [baseAccount.homeAccountId, baseAccount.environment].join(Separators.CACHE_KEY_SEPARATOR);
+        let tenantProfileAccounts: AccountInfo[] = [];
+
+        // For cases where we already have the tenant profiles preloaded like getAllAccountsMultiTenant, tenantProfiles should be passed in to avoid repeated cache access
+        if (preloadedTenantProfileAccounts) {
+            tenantProfileAccounts = preloadedTenantProfileAccounts;
         } else {
-            return null;
+            // Assuming this method is called for one account in isolation, cache access perf hit is acceptable
+            const cachedAccountKeys = this.getAccountKeys();
+            // Filter accont keys belonging to user accross tenants (including home tenant)
+            const tenantProfileAccountKeys = cachedAccountKeys.filter((key:string) => {
+                return key.indexOf(accountId) === 0;
+            });
+
+            tenantProfileAccountKeys.forEach((key: string) => {
+                const tenantProfileAccount = this.getAccount(key)?.getAccountInfo();
+                if (tenantProfileAccount) {
+                    tenantProfileAccounts.push(tenantProfileAccount);
+                }
+            });
         }
+
+        // Map tenant profile accounts to their tenantProfiles for baseAccount
+        tenantProfileAccounts.forEach((tenantProfileAccount: AccountInfo) => {
+            tenantProfiles.set(
+                tenantProfileAccount.tenantId, { 
+                    objectId: tenantProfileAccount.localAccountId,
+                    tenantId: tenantProfileAccount.tenantId,
+                    isHomeTenant: tenantProfileAccount.tenantId === tenantProfileAccount.homeAccountId.split(".")[1]
+                });
+        });
+        
+        return { ...baseAccount, tenantProfiles };
+    }
+
+    /**
+     * Returns all accounts in the cache that match the filter. If in multi-tenant mode, the accounts will be returned
+     * with their respective tenant profiles.
+     * @param accountFilter 
+     * @param multiTenantAccountsEnabled 
+     * @returns all accounts in the cache that match the filter
+     */
+    getAllAccountsFilteredBy(accountFilter: AccountFilter, multiTenantAccountsEnabled?: boolean): AccountInfo[] {
+        const fitleredAccounts: AccountInfo[] = this.getAccountsFilteredBy(accountFilter).map((accountEntity) => { return accountEntity.getAccountInfo(); });
+        if (multiTenantAccountsEnabled) {
+            return fitleredAccounts.map((accountInfo) => { return this.buildMultiTenantAccount(accountInfo); });
+        } 
+        return fitleredAccounts;
+    }
+
+    /** 
+     * Gets first AcccountInfo object that matches the provided filters
+     */
+    getAccountInfoFilteredBy(accountFilter: AccountFilter, multiTenantAccountsEnabled?: boolean): AccountInfo | null {
```

Comment:
We've now got `getAllAccountsFilteredBy`, `getAccountInfoFilteredBy` and `getAccountsFilteredBy` - what's the difference between all of these and are they all needed? Any way to converge?